### DISCUSSION TITLE
Is there a way to plot two mesh clouds on top of each other, each with a different colour?

### DISCUSSION THREAD

**Discussion by kxd195** [pyvista/pyvista#5076] [Q&A]
I've been using CloudCompare/CloudComPy in order to generate a 2.5D render of two mesh clouds together. The idea is that the difference between "cloud 1" and "cloud 2" is the difference in ground-depth that's currently being excavated. 

Is there a way in PyVista to render this in colours? Say... white for x,y,z coordinates that are the same in both, but a different colour where the two clouds are different (in my case, only the z would be different between the two files).

**Comment by banesullivan** [comment]
I'm not totally sure I follow the example, but maybe this provides a starting point for you:

```py
import pyvista as pv
import numpy as np
# Create two point clouds:
#  a) source point cloud
#  b) copy of a with a few random permutations
a = np.random.rand(100, 3)
b = a.copy()
# random mask
mask = np.random.randint(0, 2, size=a.shape).astype(bool)
# permutations
r = np.random.rand(*a.shape) * np.max(a)
b[mask] = r[mask]

# Back calculate difference
diff = a - b

ma = pv.PolyData(a)
mb = pv.PolyData(b)
mb["difference"] = diff
common = mb.threshold(0., invert=True)

pl = pv.Plotter()
pl.add_mesh(ma, color='red', render_points_as_spheres=True)
pl.add_mesh(mb, color='blue', render_points_as_spheres=True)
pl.add_mesh(common, color='white', render_points_as_spheres=True, point_size=10)
pl.show()
```
![screenshot](https://github.com/pyvista/pyvista/assets/22067021/d8d53905-733a-48aa-b8cc-c32b7296e03f)

**Reply by kxd195** [reply]
Thank you, your answer got me to about 80% of what I'm looking to achieve. Here are my current pain points:

- My data comes from xyz files containing geo-coordinates, so it's a fairly large dataset. But for plotting purposes, how can I reduce the plot points? I've looked into the Decimate filter, but it comes up with errors "Input mesh for decimation must be all triangles". When I try to triangulate using delaunay_2d, it hangs (possibly due to large dataset size).
- In my example below, the main surface area is the general surveyed ground area. The colours red-orange-yellow-green is what has been excavated in 6 hour intervals. The rest of the ground area is simply the last mesh plotted which happens to be cyan. But when you zoom in, you can see the other red-orange-yellow-green plot points as well because they are common/the same as the cyan plot. How can I go about plotting simply the "difference" between two meshes but not the common points? I've tried .boolean_intersection(), but I get the triangulation issue as above. The goal would be to plot the general area first (cyan), and then subsequent add_mesh(es) only add the intersection of the two datasets, colouring only the red/orange/yellow/green strips.

Thanks in advance. I'm coming from CloudCompare so I'm still a n00b in PyVista. 

<img width="443" alt="Screenshot 2023-10-17 at 4 12 05 PM" src="https://github.com/pyvista/pyvista/assets/35378950/47371bfe-5842-442b-b4f6-308194503576">

**Reply by banesullivan** [reply]
> how can I reduce the plot points

Do this with subsampling the points before you generate the PyVista mesh. Or if you're reading from a file and have a PyVista mesh to begin with, something like the following will do the trick:

```py
def subsample_points(mesh: pv.PolyData, subset : float = 0.02):
    """Subsample a point cloud PolyData."""
    if mesh.faces.size or mesh.lines.size:
        raise TypeError("Mesh must be a point cloud without faces.")
    ids = np.random.randint(low=0, high=mesh.n_points - 1, size=int(mesh.n_points * subset))
    return pv.PolyData(mesh.points[ids])
```

**Reply by banesullivan** [reply]
It sounds like you're treating this point cloud as a surface. You may need to mesh the surfaces from the points. PyntCloud is a great tool for this and has great interoperability with PyVista: https://github.com/daavoo/pyntcloud

**Reply by kxd195** [reply]
I tried to figure out how to use pyntcloud and able to get a pyntcloud instance from ma/mb, but the documentation seems lacking and I got nowhere further. Felt like I was getting more progress (towards my goal) using pyvista. What I think I need is getting the boolean_difference between ma and mb in your example ([source](https://docs.pyvista.org/version/stable/examples/01-filter/boolean-operations.html#boolean-difference)) but it always ends in a segmentation fault for me. Grrrrrr....

**Reply by darikg** [reply]
since you're working with pointclouds and not surfaces is it feasible to use something like a [kdtree](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.KDTree.html) to find and remove points in the cyan dataset that are within some small distance of points in the red/orange/yellow pointsets?
